limits/imap.go

```diff
@@ -27,7 +27,7 @@ func (i IMAP) CheckMailBoxCount(mailboxCount int) error {
 func (i IMAP) CheckMailBoxMessageCount(existingCount int, newCount int) error {
 	nextMessageCount := existingCount + newCount
 
-	if nextMessageCount > i.maxMessageCountPerMailbox || nextMessageCount < existingCount {
+	if uint32(nextMessageCount) > i.maxMessageCountPerMailbox || nextMessageCount < existingCount {
```

Comment:
Hmm. This brings up an interesting problem....

Since every call to `len()` in go is now going to return a 32bit singed integer (`int`) these checks will always fail and risk running into overflow territory.

We need to do 2 things here.

1)  The calculation of next message count and UID needs to be done in 64 bits.
```
nextMessageCount := int64(existingCount) + int64(count)
```

2) The comparisons also need to be promoted to int64  e.g.:
```
nextMessageCount > int64(i.maxMessageCountPerMailbox) || nextMessageCount < int64(existingCount)
```

3) Finally we need to set different limits for 32bit and 64bit compilation.
On 32bit systems the limits need to be set to `MaxInt32` since we have lot of other calculation that rely on the output of `len()` which only returns int32 values.

For 64bit, we can leave the `MaxUint32` since we can represent the calculations in the full range.

Reply:
If we're upgrading all of the `int`s to `int64`s, why do we have to use `MaxInt32` on 32 bit platforms? What's preventing the use of `MaxUint32` on 32 bit platforms if we're using 64 bit integers for all the calculations?

Reply:
P.S. forgive my ignorance throughout this PR discussion as I have no golang experience whatsoever outside of building and packaging existing code

Reply:
@benthetechguy if we can verify that`int` type on 32bits can represent a 64bit integer value, no changes need to be applied. However, based on past experiences, `int` types on 32bit system as limited to a 32bit integer only. On 64bit the `int` type is  `64bits` long.

You can check the size of the type with this snippet:
```
package main

import (
	"fmt"
	"unsafe"
)

func main() {
	fmt.Println(unsafe.Sizeof(int(0))) 
}
```

Reply:
I believe `int` on 32 bit platforms is indeed limited to the `int32` range, but you can use `int64` explicitly to get the full 64 bit range by (I think) the compiler splitting it across two instructions. Since we're making all these values `int64`s, we shouldn't have to worry about overflows, right?

Reply:
The problem is this built-in function `len() int` and other similar functions. The current snapshot state code relies on that. 

Even though we can make the database queries return the correct types, we would need to review the whole code base to ensure that that we are not losing precision range when counting things.

Reply:
If we make `MaxInt32` the limit on 32 bit platforms, wouldn't it be within the range of `int32`, preventing overflow?

Reply:
Yes.  That is correct. See my comment: https://github.com/ProtonMail/gluon/pull/383#issuecomment-1633635549

Reply:
Due to signed integer overflow: `MaxInt32 +1 = -2147483648`. On 64bit system it's enough to promote the types to int , since it is in itself already 64bits. 

On 32bits we need to do the calculations in 64bit space so that `MaxInt32+1` does not overflow.

Reply:
Okay, thank you for helping me understand.